dhall/src/Dhall.hs

```diff
@@ -791,6 +798,77 @@ list = fmap Data.Foldable.toList . sequence
 vector :: Type a -> Type (Vector a)
 vector = fmap Data.Vector.fromList . list
 
+{-| Decode a `Set` from a `List`
+
+>>> input (set natural) "[1, 2, 3]"
+fromList [1,2,3]
+
+Duplicate elements are ignored.
+-}
+set :: (Ord a) => Type a -> Type (Data.Set.Set a)
+set = fmap Data.Set.fromList . list
+
+{-| Decode a `HashSet` from a `List`
+
+>>> input (hashSet natural) "[1, 2, 3]"
+fromList [1,2,3]
+
+Duplicate elements are ignored.
+-}
+hashSet :: (Hashable a, Ord a) => Type a -> Type (Data.HashSet.HashSet a)
+hashSet = fmap Data.HashSet.fromList . list
+
+{-| Decode a `Set` from a `List` with distinct elements
+
+>>> input (setFromDistinctList natural) "[1, 2, 3]"
+fromList [1,2,3]
+
+An error is thrown if the list contains duplicates.
+
+> >>> input (setFromDistinctList natural) "[1, 1, 3]"
+> *** Exception: Error: Failed extraction
+>
+> The expression type-checked successfully but the transformation to the target
+> type failed with the following error:
+>
+> One duplicate element in the list: 1
+>
+
+> >>> input (setFromDistinctList natural) "[1, 1, 3, 3]"
+> *** Exception: Error: Failed extraction
+>
+> The expression type-checked successfully but the transformation to the target
+> type failed with the following error:
+>
+> 2 duplicates were found in the list, including 1
+>
+
+-}
+setFromDistinctList :: (Ord a, Show a) => Type a -> Type (Data.Set.Set a)
+setFromDistinctList (Type extractIn expectedIn) = Type extractOut expectedOut
+  where
+    extractOut (ListLit _ es) = case traverse extractIn es of
+        Success esSeq
+            | sameSize               -> Success esSet
+            | length duplicates == 1 -> extractError err1
+            | otherwise              -> extractError errN
+          where
+            esList = Data.Foldable.toList esSeq
+            esSet = Data.Set.fromList esList
+            sameSize = Data.Set.size esSet == Data.Sequence.length esSeq
+            duplicates = esList Data.List.\\ Data.Foldable.toList esSet
```

Comment:
If we're serious about the possibility of duplicates due to distinct Dhall terms having the same Haskell representation, this isn't sufficient. `deleteBy (\x y -> extractIn x == extractIn y)` should work.

Reply:
At this stage, `esSet` and `esList` are Haskell values already because we are inside of the `case traverse extractIn es of`. 
Maybe the `es` notation isn't helping, to be honest I don't even know what `es` stands for, I copied it from other functions.

Reply:
Ah, yes, my misunderstanding!

`es` is for "expressions", I guess. You could use `values` or `vs` for the Haskell values.